### ISSUE TITLE
Feature request: switch from C enums for tagged unions, to manifest constants

### ISSUE THREAD

**Issue by BarryBo** [FStarLang/kremlin#92]
Looking at the layout of our datastructures in-memory, I see a lot of waste caused by enums, such as the Some/None tag.  By default, C enums have sizeof(int).  But that is wasteful when there are only two options.

It would be much more efficient (both in data footprint and code size) to define FStar_Pervasives_Native_option__Prims_string_tags as uint8_t, and generate #defines for Some and None.  It gives the C compiler the ability to compact the containing struct down, if the alignment of the field that follows the tag permits.  Generally, define the storage of a tag as ln(number of tag types).

In the case of LowParse, this can be a significant win... one struct I saw has two tags - Some/None at the top-level for success/failure, and a case 'A' and case 'B' tag for two subtypes.  One subtype only stores uint8_t fields, and the other, uint16_t.  But sizeof(the type) is 10 bytes... 4 for the first tag, 4 for the second tag, then 2 for the data fields.  If the tags were uint8_t, then size would drop to 1+1+2.  

And in MSVC, this would cause the struct to match the pass-by-register / return-by-register rules, and greatly improve code quality.

**Comment by msprotz** [comment]
C++ has a feature to declare an enum type along with a type annotation (http://en.cppreference.com/w/cpp/language/enum); perhaps one more incentive to have a C++-compatible codegen at some point.

**Comment by msprotz** [comment]
(but yes your suggested change is a local one and should be cheap to implement, so it's easy to do it)

**Comment by BarryBo** [comment]
Yeah, I was about to suggest using the enum with type annotation until I realized it's a C++ feature.  So back to #define we go. :-)

**Comment by ad-l** [comment]
The C++ syntax `enum Prims_option_tag : uint8_t { Prims_None, Prims_Some }` can be used with MSVC even in C sources. For GCC you can use `-fshort-enums`

**Comment by BarryBo** [comment]
I strongly encourage you to implement this via common code... using #defines and an integral type, so the extracted code compiles with both MSVC and GCC.  If the code truly does need to differ, Kremlin should generate #if defined(_MSC_VER) / #else / #endif around the differences.

The current -windows switch is painfully fragile.  People building on cygwin forget to set it and hit errors.  People building elsewhere don't set it, and break the -windows builds.

**Comment by msprotz** [comment]
Of course -- I'm never keen on adding #ifdefs that are compiler-specific. I was just amused that this seemed to be a shared concern across compilers :). As a side note, GCC's -fshort-enums is not modular, since it forces clients of the library that "see" structs containing enums to use the same flag (or at least that's my understanding).
